File: src/Queue/Connection/Redis.php
<?php

namespace Utopia\Queue\Connection;

use Utopia\Queue\Connection;

class Redis implements Connection
{
    protected const int CONNECT_MAX_ATTEMPTS = 5;
    protected const int CONNECT_BACKOFF_MS = 100;
    protected const int CONNECT_MAX_BACKOFF_MS = 3_000;
    protected ?\Redis $redis = null;

    public function __construct(protected string $host, protected int $port = 6379, protected ?string $user = null, protected ?string $password = null, protected float $connectTimeout = -1, protected float $readTimeout = -1) {}

    public function rightPopLeftPushArray(string $queue, string $destination, int $timeout): array|false
    {
        $response = $this->rightPopLeftPush($queue, $destination, $timeout);

        if (!$response) {
            return false;
        }

        return json_decode($response, true);
    }
    public function rightPopLeftPush(string $queue, string $destination, int $timeout): string|false
    {
        $response = $this->call(fn(\Redis $redis): \Redis|string|false => $redis->bRPopLPush($queue, $destination, $timeout));

        if (!$response) {
            return false;
        }

        return $response;
    }
    public function rightPushArray(string $queue, array $value): bool
    {
        return (bool) $this->call(fn(\Redis $redis): int|\Redis|false => $redis->rPush($queue, json_encode($value)));
    }

    public function rightPush(string $queue, string $value): bool
    {
        return (bool) $this->call(fn(\Redis $redis): int|\Redis|false => $redis->rPush($queue, $value));
    }

    public function leftPushArray(string $queue, array $value): bool
    {
        return (bool) $this->call(fn(\Redis $redis): int|\Redis|false => $redis->lPush($queue, json_encode($value)));
    }

    public function leftPush(string $queue, string $value): bool
    {
        return (bool) $this->call(fn(\Redis $redis): int|\Redis|false => $redis->lPush($queue, $value));
    }

    public function leftPushMany(string $queue, array $payloads): bool
    {
        if ($payloads === []) {
            return true;
        }

        return (bool) $this->call(fn(\Redis $redis): int|\Redis|false => $redis->lPush($queue, ...$payloads));
    }

    public function rightPushMany(string $queue, array $payloads): bool
    {
        if ($payloads === []) {
            return true;
        }

        return (bool) $this->call(fn(\Redis $redis): int|\Redis|false => $redis->rPush($queue, ...$payloads));
    }

    /** @phpstan-impure */
    public function rightPopArray(string $queue, int $timeout): array|false
    {
        $response = $this->rightPop($queue, $timeout);

        if ($response === false) {
            return false;
        }

        return json_decode($response, true) ?? false;
    }

    public function rightPop(string $queue, int $timeout): string|false
    {
        $response = $this->call(fn(\Redis $redis): array|\Redis|false|null => $redis->brPop([$queue], $timeout));

        if (empty($response)) {
            return false;
        }

        return $response[1];
    }

    public function leftPopArray(string $queue, int $timeout): array|false
    {
        $response = $this->call(fn(\Redis $redis): \Redis|array|false|null => $redis->blPop($queue, $timeout));

        if (empty($response)) {
            return false;
        }

        return json_decode((string) $response[1], true) ?? false;
    }

    public function leftPop(string $queue, int $timeout): string|false
    {
        $response = $this->call(fn(\Redis $redis): \Redis|array|false|null => $redis->blPop($queue, $timeout));

        if (empty($response)) {
            return false;
        }

        return $response[1];
    }

    public function listRemove(string $queue, string $key): bool
    {
        return (bool) $this->call(fn(\Redis $redis): int|\Redis|false => $redis->lRem($queue, $key, 1), idempotent: true);
    }

    public function remove(string $key): bool
    {
        return (bool) $this->call(fn(\Redis $redis): int|\Redis|false => $redis->del($key), idempotent: true);
    }

    public function setArray(string $key, array $value, int $ttl = 0): bool
    {
        return $this->set($key, json_encode($value), $ttl);
    }

    public function set(string $key, string $value, int $ttl = 0): bool
    {
        if ($ttl > 0) {
            return $this->call(fn(\Redis $redis): bool|\Redis => $redis->setex($key, $ttl, $value), idempotent: true);
        }
        return $this->call(fn(\Redis $redis): \Redis|string|bool => $redis->set($key, $value), idempotent: true);
    }

    public function get(string $key): array|string|null
    {
        return $this->call(fn(\Redis $redis): mixed => $redis->get($key), idempotent: true);
    }

    public function listSize(string $key): int
    {
        return $this->call(fn(\Redis $redis): int|\Redis|false => $redis->lLen($key), idempotent: true);
    }

    public function increment(string $key): int
    {
        return $this->call(fn(\Redis $redis): int|\Redis|false => $redis->incr($key));
    }

    public function decrement(string $key): int
    {
        return $this->call(fn(\Redis $redis): int|\Redis|false => $redis->decr($key));
    }

    public function listRange(string $key, int $total, int $offset): array
    {
        $start = $offset;
        $end = $start + $total - 1;

        return $this->call(fn(\Redis $redis): array|\Redis|false => $redis->lRange($key, $start, $end), idempotent: true);
    }

    public function ping(): bool
    {
        try {
            $this->getRedis()->ping();

            return true;
        } catch (\Exception) {
            return false;
        }
    }

    public function close(): void
    {
        try {
            $this->redis?->close();
        } catch (\Throwable) {
        } finally {
            $this->redis = null;
        }
    }

    /**
     * Run a command on the cached client. When phpredis reports the socket
     * is gone, drop the client so the next command opens a fresh one, and
     * for a command that is safe to repeat run it again right away.
     *
     * phpredis reconnects a dropped socket by itself, but when the peer stays
     * away past its retry budget it parks the client in a failed state for
     * good, and every later command throws "went away". A worker that lived
     * through a broker failover longer than those retries would otherwise fail
     * every ack until the process restarts.
     *
     * Commands whose execution the server may already have applied (pops,
     * pushes, counters) are never replayed: the caller sees the error and the
     * broker's own retry path decides.
     *
     * @template T
     * @param callable(\Redis): T $command
     * @return T
     */
    protected function call(callable $command, bool $idempotent = false): mixed
    {
        try {
            return $command($this->getRedis());
        } catch (\RedisException $e) {
            if (!$this->isTransportError($e)) {
                throw $e;
            }

            $this->close();

            if (!$idempotent) {
                throw $e;
            }

            return $command($this->getRedis());
        }
    }

    /**
     * phpredis errors that mean the socket is gone rather than the command
     * being wrong. Anything else (WRONGTYPE, OOM, ...) is the caller's to see.
     */
    private function isTransportError(\RedisException $e): bool
    {
        return (bool) preg_match(
            '/went away|Connection (lost|closed|refused)|read error on connection/',
            $e->getMessage(),
        );
    }

    protected function getRedis(): \Redis
    {
        if ($this->redis instanceof \Redis) {
            return $this->redis;
        }

        $connectTimeout = $this->connectTimeout < 0 ? 0 : $this->connectTimeout;

        for ($attempt = 1; $attempt <= self::CONNECT_MAX_ATTEMPTS; $attempt++) {
            $redis = new \Redis();

            try {
                $redis->connect($this->host, $this->port, $connectTimeout);

                if ($this->password !== null && $this->password !== '') {
                    $hasUser = $this->user !== null && $this->user !== '';
                    $redis->auth($hasUser ? [$this->user, $this->password] : $this->password);
                }

                if ($this->readTimeout >= 0) {
                    $redis->setOption(\Redis::OPT_READ_TIMEOUT, $this->readTimeout);
                }

                $this->redis = $redis;
                return $this->redis;
            } catch (\RedisException $e) {
                try {
                    $redis->close();
                } catch (\Throwable) {
                }

                if ($attempt === self::CONNECT_MAX_ATTEMPTS) {
                    throw new \RedisException(
                        \sprintf(
                            'Failed to connect to Redis at %s:%d after %d attempts: %s',
                            $this->host,
                            $this->port,
                            self::CONNECT_MAX_ATTEMPTS,
                            $e->getMessage(),
                        ),
                        $e->getCode(),
                        $e,
                    );
                }

                // Exponential backoff with full jitter to avoid thundering herd on recovery.
                $backoffMs = min(
                    self::CONNECT_MAX_BACKOFF_MS,
                    self::CONNECT_BACKOFF_MS * (2 ** ($attempt - 1)),
                );
                usleep(mt_rand(0, $backoffMs) * 1000);
            }
        }

        throw new \RedisException(\sprintf(
            'Unreachable: Redis connect loop for %s:%d exited after %d attempts without success or exception.',
            $this->host,
            $this->port,
            self::CONNECT_MAX_ATTEMPTS,
        ));
    }
}

File: 3.x.x/crates-utopia/queue/src/connection/redis.rs
use std::thread;
use std::time::Duration;

use parking_lot::Mutex;
use redis::{Commands, RedisError};
use serde_json::Value;

use super::{Connection, StoredValue};
use crate::error::QueueError;

const CONNECT_MAX_ATTEMPTS: u32 = 5;
const CONNECT_BACKOFF_MS: u64 = 100;
const CONNECT_MAX_BACKOFF_MS: u64 = 3_000;

/// Single-node Redis connection.
///
/// PHP `Utopia\Queue\Connection\Redis`.
pub struct Redis {
    host: String,
    port: u16,
    user: Option<String>,
    password: Option<String>,
    connect_timeout: f64,
    read_timeout: f64,
    conn: Mutex<Option<redis::Connection>>,
}

impl std::fmt::Debug for Redis {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Redis")
            .field("host", &self.host)
            .field("port", &self.port)
            .finish_non_exhaustive()
    }
}

impl Redis {
    pub fn new(host: impl Into<String>, port: u16) -> Self {
        Self {
            host: host.into(),
            port,
            user: None,
            password: None,
            connect_timeout: -1.0,
            read_timeout: -1.0,
            conn: Mutex::new(None),
        }
    }

    pub fn with_auth(mut self, user: Option<String>, password: Option<String>) -> Self {
        self.user = user;
        self.password = password;
        self
    }

    pub fn with_timeouts(mut self, connect_timeout: f64, read_timeout: f64) -> Self {
        self.connect_timeout = connect_timeout;
        self.read_timeout = read_timeout;
        self
    }

    fn url(&self) -> String {
        match (&self.user, &self.password) {
            (Some(user), Some(pass)) => {
                format!("redis://{user}:{pass}@{}:{}/", self.host, self.port)
            }
            (None, Some(pass)) => format!("redis://:{pass}@{}:{}/", self.host, self.port),
            _ => format!("redis://{}:{}/", self.host, self.port),
        }
    }

    fn connect_once(&self) -> Result<redis::Connection, RedisError> {
        let client = redis::Client::open(self.url().as_str())?;
        let conn = if self.connect_timeout >= 0.0 {
            client.get_connection_with_timeout(Duration::from_secs_f64(self.connect_timeout))?
        } else {
            client.get_connection()?
        };
        if self.read_timeout >= 0.0 {
            let _ = conn.set_read_timeout(Some(Duration::from_secs_f64(self.read_timeout)));
        }
        Ok(conn)
    }

    fn with_conn<R>(
        &self,
        f: impl FnOnce(&mut redis::Connection) -> Result<R, RedisError>,
    ) -> Result<R, QueueError> {
        {
            let mut guard = self.conn.lock();
            if let Some(conn) = guard.as_mut() {
                return f(conn).map_err(redis_err);
            }
        }
        let mut last_err: Option<RedisError> = None;
        for attempt in 1..=CONNECT_MAX_ATTEMPTS {
            match self.connect_once() {
                Ok(conn) => {
                    let mut guard = self.conn.lock();
                    *guard = Some(conn);
                    return f(guard.as_mut().expect("just inserted")).map_err(redis_err);
                }
                Err(e) => {
                    last_err = Some(e);
                    if attempt == CONNECT_MAX_ATTEMPTS {
                        break;
                    }
                    let backoff_ms = CONNECT_MAX_BACKOFF_MS
                        .min(CONNECT_BACKOFF_MS.saturating_mul(1u64 << (attempt - 1)));
                    let jitter = fastrand_u64(backoff_ms);
                    thread::sleep(Duration::from_millis(jitter));
                }
            }
        }
        Err(QueueError::redis(format!(
            "Failed to connect to Redis at {}:{} after {CONNECT_MAX_ATTEMPTS} attempts: {}",
            self.host,
            self.port,
            last_err
                .map(|e| e.to_string())
                .unwrap_or_else(|| "unknown".into()),
        )))
    }
}

impl Connection for Redis {
    fn right_push_array(&self, queue: &str, payload: &Value) -> Result<bool, QueueError> {
        let encoded =
            serde_json::to_string(payload).map_err(|e| QueueError::Other(e.to_string()))?;
        self.right_push(queue, &encoded)
    }

    fn right_pop_array(&self, queue: &str, timeout: i64) -> Result<Option<Value>, QueueError> {
        match self.right_pop(queue, timeout)? {
            Some(raw) => Ok(serde_json::from_str(&raw).ok()),
            None => Ok(None),
        }
    }

    fn right_pop_left_push_array(
        &self,
        queue: &str,
        destination: &str,
        timeout: i64,
    ) -> Result<Option<Value>, QueueError> {
        match self.right_pop_left_push(queue, destination, timeout)? {
            Some(raw) => Ok(serde_json::from_str(&raw).ok()),
            None => Ok(None),
        }
    }

    fn left_push_array(&self, queue: &str, payload: &Value) -> Result<bool, QueueError> {
        let encoded =
            serde_json::to_string(payload).map_err(|e| QueueError::Other(e.to_string()))?;
        self.left_push(queue, &encoded)
    }

    fn left_pop_array(&self, queue: &str, timeout: i64) -> Result<Option<Value>, QueueError> {
        self.with_conn(|conn| {
            let result: Option<(String, String)> = conn.blpop(queue, timeout as f64)?;
            Ok(result.and_then(|(_, v)| serde_json::from_str(&v).ok()))
        })
    }

    fn right_push(&self, queue: &str, payload: &str) -> Result<bool, QueueError> {
        self.with_conn(|conn| {
            let n: i64 = conn.rpush(queue, payload)?;
            Ok(n > 0)
        })
    }

    fn right_pop(&self, queue: &str, timeout: i64) -> Result<Option<String>, QueueError> {
        self.with_conn(|conn| {
            let result: Option<(String, String)> = conn.brpop(queue, timeout as f64)?;
            Ok(result.map(|(_, v)| v))
        })
    }

    fn right_pop_left_push(
        &self,
        queue: &str,
        destination: &str,
        timeout: i64,
    ) -> Result<Option<String>, QueueError> {
        self.with_conn(|conn| {
            let result: Option<String> = conn.brpoplpush(queue, destination, timeout as f64)?;
            Ok(result)
        })
    }

    fn left_push(&self, queue: &str, payload: &str) -> Result<bool, QueueError> {
        self.with_conn(|conn| {
            let n: i64 = conn.lpush(queue, payload)?;
            Ok(n > 0)
        })
    }

    fn left_pop(&self, queue: &str, timeout: i64) -> Result<Option<String>, QueueError> {
        self.with_conn(|conn| {
            let result: Option<(String, String)> = conn.blpop(queue, timeout as f64)?;
            Ok(result.map(|(_, v)| v))
        })
    }

    fn list_remove(&self, queue: &str, key: &str) -> Result<bool, QueueError> {
        self.with_conn(|conn| {
            let n: i64 = conn.lrem(queue, 1, key)?;
            Ok(n > 0)
        })
    }

    fn list_size(&self, key: &str) -> Result<i64, QueueError> {
        self.with_conn(|conn| conn.llen(key))
    }

    fn list_range(&self, key: &str, total: i64, offset: i64) -> Result<Vec<Value>, QueueError> {
        let start = offset;
        let end = start + total - 1;
        self.with_conn(|conn| {
            let raw: Vec<String> = conn.lrange(key, start as isize, end as isize)?;
            Ok(raw
                .into_iter()
                .map(|s| serde_json::from_str(&s).unwrap_or(Value::String(s)))
                .collect())
        })
    }

    fn remove(&self, key: &str) -> Result<bool, QueueError> {
        self.with_conn(|conn| {
            let n: i64 = conn.del(key)?;
            Ok(n > 0)
        })
    }

    fn set(&self, key: &str, value: &str, ttl: i64) -> Result<bool, QueueError> {
        self.with_conn(|conn| {
            if ttl > 0 {
                conn.set_ex(key, value, ttl as u64)
            } else {
                conn.set(key, value)
            }
        })
    }

    fn get(&self, key: &str) -> Result<Option<StoredValue>, QueueError> {
        self.with_conn(|conn| {
            let raw: Option<String> = conn.get(key)?;
            Ok(raw.map(StoredValue::String))
        })
    }

    fn set_array(&self, key: &str, value: &Value, ttl: i64) -> Result<bool, QueueError> {
        let encoded = serde_json::to_string(value).map_err(|e| QueueError::Other(e.to_string()))?;
        self.set(key, &encoded, ttl)
    }

    fn increment(&self, key: &str) -> Result<i64, QueueError> {
        self.with_conn(|conn| conn.incr(key, 1i64))
    }

    fn decrement(&self, key: &str) -> Result<i64, QueueError> {
        self.with_conn(|conn| conn.decr(key, 1i64))
    }

    fn ping(&self) -> Result<bool, QueueError> {
        match self.with_conn(|conn| {
            let _: String = redis::cmd("PING").query(conn)?;
            Ok(())
        }) {
            Ok(()) => Ok(true),
            Err(_) => Ok(false),
        }
    }

    fn close(&self) {
        *self.conn.lock() = None;
    }
}

pub(crate) fn redis_err(err: RedisError) -> QueueError {
    QueueError::redis(err.to_string())
}

fn fastrand_u64(max_inclusive: u64) -> u64 {
    if max_inclusive == 0 {
        return 0;
    }
    use std::time::{SystemTime, UNIX_EPOCH};
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.subsec_nanos() as u64)
        .unwrap_or(1);
    nanos % (max_inclusive + 1)
}
